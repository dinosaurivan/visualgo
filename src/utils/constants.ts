export const DEFAULT_ARRAY_SIZE = 5;
export const DEFAULT_QUEUE_SIZE = 7;

export enum Direction {
  Ascending = "ascending",
  Descending = "descending",
};

export enum ElementColors {
  Default = "default",
  Selected = "selected",
  Changing = "changing",
  Modified = "modified",
  Invisible = "invisible",
};

export enum SortingAlgorithms {
  Bubble = "bubble",
  Selection = "selection",
};

export enum ElementCaptions {
  Top = "top",
  Head = "head",
  Tail = "tail",
};

export enum StackActions {
  Pop = "pop",
  Push = "push",
  Clear = "clear",
};

export enum QueueActions {
  Enqueue = "enqueue",
  Dequeue = "dequeue",
  Clear = "clear",
};

export enum LinkedListActions {
  Pop = "pop",
  Push = "push",
  Shift = "shift",
  Insert = "insert",
  Remove = "remove",
  Unshift = "unshift",
};

export enum Delay {
  None = 0,
  Medium = 1000,
};
