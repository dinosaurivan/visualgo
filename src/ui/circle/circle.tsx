import React from "react";
import styles from "./circle.module.css";
import { ElementColors } from "../../utils/constants";

interface CircleProps {
  color?: ElementColors;
  value?: string;
  above?: string | React.ReactElement | null;
  index?: number;
  below?: string | React.ReactElement | null;
  extraClass?: string;
  isSmall?: boolean;
}

export const Circle: React.FC<CircleProps> = ({
  color = ElementColors.Default,
  value,
  above,
  index,
  below,
  extraClass = "",
  isSmall,
}) => {
  return (
    <div 
      data-testid="circle"
      className={`${styles.content} ${extraClass}`} 
    >
      <div
        data-testid="above"
        className={`text text_type_input text_color_input mb-4 ${
          styles.above
        } ${
          styles.absolute
        } ${
          styles[typeof above === "string" ? "string" : "element"]
        }`}
      >
        {above}
      </div>
      <div
        data-testid="inside"
        className={`${styles.circle}  ${isSmall ? styles.small : ""} ${
          styles[color]
        }`}
      >
        <p
          className={`text text_type_circle text_color_input ${styles.value}`}
        >
          {value}
        </p>
      </div>
      <p
        data-testid="index"
        className={`text text_type_input text_color_input mt-4 ${styles.absolute} ${styles.index}`}
      >
        {index?.toString()}
      </p>
      <div
        data-testid="below"
        className={`text text_type_input text_color_input mt-4 ${
          index?.toString() ? styles.below60 : styles.below30
        } ${
          styles.absolute
        } ${
          styles[typeof below === "string" ? "string" : "element"]
        }`}
      >
        {below}
      </div>
    </div>
  );
};
