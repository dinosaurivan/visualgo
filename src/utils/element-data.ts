// utils 
import { ElementColors } from "./constants";



export class ElementData<T> {
  value: T;
  color: ElementColors;
  isHead: boolean;
  isTail: boolean;  
  valueAbove?: T;
  valueBelow?: T;
  
  constructor(
    value: T, 
    color: ElementColors = ElementColors.Default,
    isHead: boolean = false,
    isTail: boolean = false,      
    valueAbove?: T, 
    valueBelow?: T, 
  ) {
    this.value = value;
    this.color = color;
    this.isHead = isHead;
    this.isTail = isTail;
    this.valueAbove = valueAbove;
    this.valueBelow = valueBelow;
  };
};
