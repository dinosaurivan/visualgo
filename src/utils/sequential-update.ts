// libraries 
import { Dispatch, SetStateAction } from "react";

// utils 
import sleep from "./sleep";
import { Delay } from "./constants";
import ElementData from "./element-data";



const sequentialUpdate = async <T>(
  
  steps: Array<Array<ElementData<T>>>, 
  stepSetter: Dispatch<SetStateAction<Array<ElementData<T>>>>,
  inProgressSetter: Dispatch<SetStateAction<boolean>>,
  componentIsMounted: () => boolean,
  
) => {
  
  let isFirstIteration = true;
  
  for (const step of steps) {
    
    await sleep(isFirstIteration ? Delay.None : Delay.Medium);
    isFirstIteration = false;
    
    if (componentIsMounted()) {
      stepSetter(step);
    };
  };
  
  inProgressSetter(false);
};

export default sequentialUpdate;
