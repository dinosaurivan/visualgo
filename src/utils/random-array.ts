import { ElementData } from "./element-data";
import { DEFAULT_ARRAY_SIZE, MAX_ARRAY_VALUE, MIN_ARRAY_VALUE } from "./constants";



export const randomNumbersArray = (size?: number) => {
  const result = Array.from(
    { length: size || DEFAULT_ARRAY_SIZE }, 
    () => new ElementData<number>(
      Math.floor(MIN_ARRAY_VALUE + Math.random() * MAX_ARRAY_VALUE)
    )
  );
  result[0].isHead = true;
  result[result.length-1].isTail = true;
  return result;
};



export const randomStringsArray = (size?: number) => {
  const result = Array.from(
    { length: size || DEFAULT_ARRAY_SIZE }, 
    () => new ElementData<string>(
      String(
        Math.floor(MIN_ARRAY_VALUE + Math.random() * MAX_ARRAY_VALUE)
      )
    )
  );
  result[0].isHead = true;
  result[result.length-1].isTail = true;
  return result;
};
