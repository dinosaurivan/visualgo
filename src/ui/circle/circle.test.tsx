// libraries 
import React from "react";
import { render, screen } from "@testing-library/react";

// components
import { Circle } from "./circle";
import { ElementColors } from "../../utils/constants";



describe(
  "Circle UI component tests", 
  () => {
    
    test(
      "renders correctly with value provided", 
      () => {
        // arrange & act
        render(<Circle value="value inside" />);
        const component = screen.getByTestId("circle");
        const circle = screen.getByTestId("inside");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(circle).toBeInTheDocument();
        expect(circle).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(circle).toHaveClass("circle default");
        expect(circle).toHaveTextContent("value inside");
      }
    );
    
    test(
      "renders correctly without value provided", 
      () => {
        // arrange & act
        render(<Circle />);
        const component = screen.getByTestId("circle");
        const circle = screen.getByTestId("inside");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(circle).toBeInTheDocument();
        expect(circle).toMatchSnapshot();        
        
        expect(component).toHaveClass("content");
        expect(circle).toHaveClass("circle default");
        expect(circle).toHaveTextContent("");
      }
    );
    
    test(
      "renders correctly with isSmall prop is set to true",       
      () => {
        // arrange & act
        render(<Circle isSmall={true} value="value inside" />);
        const component = screen.getByTestId("circle");
        const circle = screen.getByTestId("inside");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(circle).toBeInTheDocument();
        expect(circle).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(circle).toHaveClass("circle default small");
        expect(circle).toHaveTextContent("value inside");      
    });       
    
    test(
      "renders correctly with color prop is set to changing", 
      () => {
        // arrange & act
        render(<Circle value="value inside" color={ElementColors.Changing} />);
        const component = screen.getByTestId("circle");
        const circle = screen.getByTestId("inside");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(circle).toBeInTheDocument();
        expect(circle).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(circle).toHaveClass("circle changing");
        expect(circle).toHaveTextContent("value inside");
      }
    );    
    
    test(
      "renders correctly with color prop is set to modified", 
      () => {
        // arrange & act
        render(<Circle value="value inside" color={ElementColors.Modified} />);
        const component = screen.getByTestId("circle");
        const circle = screen.getByTestId("inside");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(circle).toBeInTheDocument();
        expect(circle).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(circle).toHaveClass("circle modified");
        expect(circle).toHaveTextContent("value inside");
      }
    );       
    
    test(
      "renders correctly with string above",       
      () => {
        // arrange & act
        render(<Circle above="value above" />);
        const component = screen.getByTestId("circle");
        const above = screen.getByTestId("above");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(above).toBeInTheDocument();
        expect(above).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(above).toHaveClass("above absolute string");
        expect(above).toHaveTextContent("value above");       
      }
    );    
    
    test(
      "renders correctly with string below",       
      () => {
        // arrange & act
        render(<Circle below="value below" />);
        const component = screen.getByTestId("circle");
        const below = screen.getByTestId("below");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(below).toBeInTheDocument();
        expect(below).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(below).toHaveClass("below30 absolute string");
        expect(below).toHaveTextContent("value below");       
      }
    );       
    
    test(
      "renders correctly with element above",       
      () => {
        // arrange & act
        render(<Circle above={<Circle value="value inside" />} />);
        const component = screen.getAllByTestId("circle")[0];
        const above = screen.getAllByTestId("above")[0];
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(above).toBeInTheDocument();
        expect(above).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(above).toHaveClass("above absolute element");
        expect(above).toHaveTextContent("value inside");      
      }
    );    
    
    test(
      "renders correctly with element below",       
      () => {
        // arrange & act
        render(<Circle below={<Circle value="value inside" />} />);
        const component = screen.getAllByTestId("circle")[0];
        const below = screen.getAllByTestId("below")[0];
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(below).toBeInTheDocument();
        expect(below).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(below).toHaveClass("below30 absolute element");
        expect(below).toHaveTextContent("value inside");      
      }
    );        
    
    test(
      "renders correctly with an index",       
      () => {
        // arrange & act
        render(<Circle index={42} />);
        const component = screen.getByTestId("circle");
        const index = screen.getByTestId("index");
        
        // assert
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
        expect(index).toBeInTheDocument();
        expect(index).toMatchSnapshot();
        
        expect(component).toHaveClass("content");
        expect(index).toHaveClass("index absolute");
        expect(index).toHaveTextContent("42");      
      }
    );       
  }
);
