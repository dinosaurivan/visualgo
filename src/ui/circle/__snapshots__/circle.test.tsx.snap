// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle UI component tests renders correctly with an index 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  >
    42
  </p>
  <div
    class="text text_type_input text_color_input mt-4 below60 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with an index 2`] = `
<p
  class="text text_type_input text_color_input mt-4 absolute index"
  data-testid="index"
>
  42
</p>
`;

exports[`Circle UI component tests renders correctly with color prop is set to changing 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   changing"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    >
      value inside
    </p>
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with color prop is set to changing 2`] = `
<div
  class="circle   changing"
  data-testid="inside"
>
  <p
    class="text text_type_circle text_color_input value"
  >
    value inside
  </p>
</div>
`;

exports[`Circle UI component tests renders correctly with color prop is set to modified 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   modified"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    >
      value inside
    </p>
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with color prop is set to modified 2`] = `
<div
  class="circle   modified"
  data-testid="inside"
>
  <p
    class="text text_type_circle text_color_input value"
  >
    value inside
  </p>
</div>
`;

exports[`Circle UI component tests renders correctly with element above 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  >
    <div
      class="content "
      data-testid="circle"
    >
      <div
        class="text text_type_input text_color_input mb-4 above absolute element"
        data-testid="above"
      />
      <div
        class="circle   default"
        data-testid="inside"
      >
        <p
          class="text text_type_circle text_color_input value"
        >
          value inside
        </p>
      </div>
      <p
        class="text text_type_input text_color_input mt-4 absolute index"
        data-testid="index"
      />
      <div
        class="text text_type_input text_color_input mt-4 below30 absolute element"
        data-testid="below"
      />
    </div>
  </div>
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with element above 2`] = `
<div
  class="text text_type_input text_color_input mb-4 above absolute element"
  data-testid="above"
>
  <div
    class="content "
    data-testid="circle"
  >
    <div
      class="text text_type_input text_color_input mb-4 above absolute element"
      data-testid="above"
    />
    <div
      class="circle   default"
      data-testid="inside"
    >
      <p
        class="text text_type_circle text_color_input value"
      >
        value inside
      </p>
    </div>
    <p
      class="text text_type_input text_color_input mt-4 absolute index"
      data-testid="index"
    />
    <div
      class="text text_type_input text_color_input mt-4 below30 absolute element"
      data-testid="below"
    />
  </div>
</div>
`;

exports[`Circle UI component tests renders correctly with element below 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  >
    <div
      class="content "
      data-testid="circle"
    >
      <div
        class="text text_type_input text_color_input mb-4 above absolute element"
        data-testid="above"
      />
      <div
        class="circle   default"
        data-testid="inside"
      >
        <p
          class="text text_type_circle text_color_input value"
        >
          value inside
        </p>
      </div>
      <p
        class="text text_type_input text_color_input mt-4 absolute index"
        data-testid="index"
      />
      <div
        class="text text_type_input text_color_input mt-4 below30 absolute element"
        data-testid="below"
      />
    </div>
  </div>
</div>
`;

exports[`Circle UI component tests renders correctly with element below 2`] = `
<div
  class="text text_type_input text_color_input mt-4 below30 absolute element"
  data-testid="below"
>
  <div
    class="content "
    data-testid="circle"
  >
    <div
      class="text text_type_input text_color_input mb-4 above absolute element"
      data-testid="above"
    />
    <div
      class="circle   default"
      data-testid="inside"
    >
      <p
        class="text text_type_circle text_color_input value"
      >
        value inside
      </p>
    </div>
    <p
      class="text text_type_input text_color_input mt-4 absolute index"
      data-testid="index"
    />
    <div
      class="text text_type_input text_color_input mt-4 below30 absolute element"
      data-testid="below"
    />
  </div>
</div>
`;

exports[`Circle UI component tests renders correctly with isSmall prop is set to true 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle  small default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    >
      value inside
    </p>
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with isSmall prop is set to true 2`] = `
<div
  class="circle  small default"
  data-testid="inside"
>
  <p
    class="text text_type_circle text_color_input value"
  >
    value inside
  </p>
</div>
`;

exports[`Circle UI component tests renders correctly with string above 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute string"
    data-testid="above"
  >
    value above
  </div>
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with string above 2`] = `
<div
  class="text text_type_input text_color_input mb-4 above absolute string"
  data-testid="above"
>
  value above
</div>
`;

exports[`Circle UI component tests renders correctly with string below 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute string"
    data-testid="below"
  >
    value below
  </div>
</div>
`;

exports[`Circle UI component tests renders correctly with string below 2`] = `
<div
  class="text text_type_input text_color_input mt-4 below30 absolute string"
  data-testid="below"
>
  value below
</div>
`;

exports[`Circle UI component tests renders correctly with value provided 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    >
      value inside
    </p>
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly with value provided 2`] = `
<div
  class="circle   default"
  data-testid="inside"
>
  <p
    class="text text_type_circle text_color_input value"
  >
    value inside
  </p>
</div>
`;

exports[`Circle UI component tests renders correctly without value provided 1`] = `
<div
  class="content "
  data-testid="circle"
>
  <div
    class="text text_type_input text_color_input mb-4 above absolute element"
    data-testid="above"
  />
  <div
    class="circle   default"
    data-testid="inside"
  >
    <p
      class="text text_type_circle text_color_input value"
    />
  </div>
  <p
    class="text text_type_input text_color_input mt-4 absolute index"
    data-testid="index"
  />
  <div
    class="text text_type_input text_color_input mt-4 below30 absolute element"
    data-testid="below"
  />
</div>
`;

exports[`Circle UI component tests renders correctly without value provided 2`] = `
<div
  class="circle   default"
  data-testid="inside"
>
  <p
    class="text text_type_circle text_color_input value"
  />
</div>
`;
